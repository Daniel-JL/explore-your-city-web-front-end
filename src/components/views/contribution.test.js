import React, { useState, useEffect, useRef } from 'react';
import {
  render, screen, waitFor, unmountComponentAtNode, act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
// import { act } from 'react-dom/test-utils';
// import mapboxgl from 'mapbox-gl/dist/mapbox-gl';

import userEvent from '@testing-library/user-event';
import {
  Router,
} from 'react-router-dom';
import { createMemoryHistory } from 'history';
import fetch from 'jest-fetch-mock';
import '@testing-library/jest-dom/extend-expect';
// import Enzyme, { shallow, mount, render } from 'enzyme';
// import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
// import Map from '../stories/map';
// import { removeMapMarkerFromList } from '../stories/map';
import Contribution from './contribution';
import mapboxgl from 'mapbox-gl';
// Enzyme.configure({ adapter: new Adapter() });

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
  GeolocateControl: () => {},
  Marker: () => ({
    addTo: () => {},
    setLngLat: () => {},
    remove: () => {},
  }),
  Map: () => ({
    addControl: () => {},
    on: () => {},
    remove: () => {},
    getZoom: () => {},
  }),
  NavigationControl: () => {},
}));

const mockCallback = jest.fn();

beforeEach(() => {
  mockCallback.mockClear();
  fetch.resetMocks();
});

describe('Contribution page', () => {
  it('should render without throwing an error', () => {
    const ContributionComponent = render(
      <Contribution />
      ,
    );
    expect(ContributionComponent).toMatchSnapshot();
  });

  it('should update story box when new item is added', async () => {
    const { getByRole, rerender } = render(
      <Contribution setModuleCallback={mockCallback} />
      ,
    );

    const addLocationButton = screen.getByRole('button', { name: /add-location-button/i });
    const addressBox = screen.getByRole('textbox', { name: /address-field/i });
    const storyBox = screen.getByRole('textbox', { name: /location-story/i });
    const addressSearchBox = screen.getByRole('textbox', { name: /Search/i });
    const addressListTextArea = screen.getByRole('textbox', { name: /address-list-text-area/i });

    userEvent.type(addressSearchBox, 'Choriner Höfe, Berlin');
    userEvent.type(addressListTextArea, 'A cool story');
    userEvent.click(addLocationButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('Choriner Höfe, Berlin');
      expect(storyBox.value).toBe('A cool story');
    });

    userEvent.clear(addressSearchBox);
    userEvent.type(addressSearchBox, 'Choriner Straße, Berlin');
    userEvent.type(addressListTextArea, 'Another cool story');
    userEvent.click(addLocationButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('Choriner Straße, Berlin');
      expect(storyBox.value).toBe('Another cool story');
    });
    expect(mockCallback).toHaveBeenCalledTimes(2);
    // expect(mapboxgl.Marker.addTo).toHaveBeenCalledTimes(2);
  });

  it('should delete BrowserListItem item when delete button is clicked', async () => {
    const { getByRole, rerender } = render(
      <Contribution setModuleCallback={mockCallback} />
      ,
    );

    const addLocationButton = screen.getByRole('button', { name: /add-location-button/i });
    const trashButton = screen.getByRole('button', { name: /trash-button/i });
    const addressBox = screen.getByRole('textbox', { name: /address-field/i });
    const storyBox = screen.getByRole('textbox', { name: /location-story/i });
    const addressSearchBox = screen.getByRole('textbox', { name: /Search/i });
    const addressListTextArea = screen.getByRole('textbox', { name: /address-list-text-area/i });

    userEvent.type(addressSearchBox, 'Choriner Höfe, Berlin');
    userEvent.type(addressListTextArea, 'A cool story');
    userEvent.click(addLocationButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('Choriner Höfe, Berlin');
      expect(storyBox.value).toBe('A cool story');
    });

    userEvent.clear(addressSearchBox);
    userEvent.type(addressSearchBox, 'Choriner Straße, Berlin');
    userEvent.type(addressListTextArea, 'Another cool story');
    userEvent.click(addLocationButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('Choriner Straße, Berlin');
      expect(storyBox.value).toBe('Another cool story');
    });

    userEvent.click(trashButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('Choriner Höfe, Berlin');
      expect(storyBox.value).toBe('A cool story');
    });

    userEvent.click(trashButton);

    await waitFor(() => {
      expect(addressBox.value).toBe('');
      expect(storyBox.value).toBe('');
    });

    expect(mockCallback).toHaveBeenCalledTimes(4);
  });

  // it('should update map when new item is added', async () => {
  //   const { getByRole, rerender } = render(
  //     <Contribution />
  //     ,
  //   );

  //   const addLocationButton = screen.getByRole('button', { name: /add-location-button/i });
  //   const addressSearchBox = screen.getByRole('textbox', { name: /Search/i });
  //   const addressListTextArea = screen.getByRole('textbox', { name: /address-list-text-area/i });

  //   userEvent.type(addressSearchBox, 'Choriner Höfe, Berlin');
  //   userEvent.type(addressListTextArea, 'A cool story');
  //   userEvent.click(addLocationButton);

  //   const markers = screen.findAllByLabelText('Map marker');

  //   await waitFor(() => {
  //     expect(markers).toHaveLength(1);
  //   });
  //   // userEvent.clear(addressSearchBox);
  //   // userEvent.type(addressSearchBox, 'Choriner Straße, Berlin');
  //   // userEvent.type(addressListTextArea, 'Another cool story');
  //   // userEvent.click(addLocationButton);

  //   // await waitFor(() => {
  //   //   expect(addressBox.value).toBe('Choriner Straße, Berlin');
  //   //   expect(storyBox.value).toBe('Another cool story');
  //   // });
  // });
});
