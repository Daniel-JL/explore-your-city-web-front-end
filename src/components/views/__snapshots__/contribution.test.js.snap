// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contribution page should render without throwing an error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-fFubgz jfAIee"
      >
        <div
          class="sc-bkzZxe bzMjUo"
        >
          <div
            class="sc-idOhPF dyCZxh"
            id="addressListContainer"
          >
            <div
              class="sc-dQppl jjFnjm"
              id="address-list-container"
            >
              <div
                class="sc-bqyKva hKNAsX"
              >
                <div
                  class="sc-kstrdz hrtAwC"
                  id="geocoder-container"
                >
                  <div
                    class="mapboxgl-ctrl-geocoder mapboxgl-ctrl"
                  >
                    <svg
                      class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-search"
                      height="18"
                      viewBox="0 0 18 18"
                      width="18"
                      xml:space="preserve"
                    >
                      <path
                        d="M7.4 2.5c-2.7 0-4.9 2.2-4.9 4.9s2.2 4.9 4.9 4.9c1 0 1.8-.2 2.5-.8l3.7 3.7c.2.2.4.3.8.3.7 0 1.1-.4 1.1-1.1 0-.3-.1-.5-.3-.8L11.4 10c.4-.8.8-1.6.8-2.5.1-2.8-2.1-5-4.8-5zm0 1.6c1.8 0 3.2 1.4 3.2 3.2s-1.4 3.2-3.2 3.2-3.3-1.3-3.3-3.1 1.4-3.3 3.3-3.3z"
                      />
                    </svg>
                    <input
                      aria-label="Search"
                      class="mapboxgl-ctrl-geocoder--input"
                      placeholder="Search"
                      type="text"
                    />
                    <div
                      class="suggestions-wrapper"
                    >
                      <ul
                        class="suggestions"
                        style="display: none;"
                      />
                    </div>
                    <div
                      class="mapboxgl-ctrl-geocoder--pin-right"
                    >
                      <button
                        aria-label="Clear"
                        class="mapboxgl-ctrl-geocoder--button"
                      >
                        <svg
                          class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-close"
                          height="18"
                          viewBox="0 0 18 18"
                          width="18"
                          xml:space="preserve"
                        >
                          <path
                            d="M3.8 2.5c-.6 0-1.3.7-1.3 1.3 0 .3.2.7.5.8L7.2 9 3 13.2c-.3.3-.5.7-.5 1 0 .6.7 1.3 1.3 1.3.3 0 .7-.2 1-.5L9 10.8l4.2 4.2c.2.3.7.3 1 .3.6 0 1.3-.7 1.3-1.3 0-.3-.2-.7-.3-1l-4.4-4L15 4.6c.3-.2.5-.5.5-.8 0-.7-.7-1.3-1.3-1.3-.3 0-.7.2-1 .3L9 7.1 4.8 2.8c-.3-.1-.7-.3-1-.3z"
                          />
                        </svg>
                      </button>
                      <svg
                        class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-loading"
                        height="18"
                        viewBox="0 0 18 18"
                        width="18"
                        xml:space="preserve"
                      >
                        <path
                          d="M4.4 4.4l.8.8c2.1-2.1 5.5-2.1 7.6 0l.8-.8c-2.5-2.5-6.7-2.5-9.2 0z"
                          fill="#333"
                        />
                        <path
                          d="M12.8 12.9c-2.1 2.1-5.5 2.1-7.6 0-2.1-2.1-2.1-5.5 0-7.7l-.8-.8c-2.5 2.5-2.5 6.7 0 9.2s6.6 2.5 9.2 0 2.5-6.6 0-9.2l-.8.8c2.2 2.1 2.2 5.6 0 7.7z"
                          opacity=".1"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <br />
                <textarea
                  aria-label="address-list-text-area"
                  class="sc-hBEYos sc-fodVxV diFCsr kmUuBX"
                  id="address-list-text-area"
                  placeholder="Enter location story here"
                />
                <button
                  aria-label="add-location-button"
                  class="sc-hKgILt sc-iBPRYJ kYZczb gHMeds"
                  id="add-location-button"
                >
                  +
                </button>
              </div>
              <div
                class="sc-jrAGrp jmOtbB"
              >
                <div
                  class="sc-kEjbxe ZjlhN"
                >
                  <button
                    aria-label="left-arrow"
                    class="sc-dlfnbm ddKcEg"
                  >
                    Left
                  </button>
                  <button
                    aria-label="right-arrow"
                    class="sc-dlfnbm ddKcEg"
                  >
                    Right
                  </button>
                  <button
                    aria-label="trash-button"
                    class="sc-dlfnbm sc-eCssSg ddKcEg iQemPK"
                    id="trash-button"
                  >
                    Trash
                  </button>
                  <button
                    aria-label="hide-text-button"
                    class="sc-dlfnbm ddKcEg"
                    id="hide-text-button"
                  >
                    Hide/Show Text
                  </button>
                </div>
                <input
                  aria-label="address-field"
                  class="sc-iqHYGH bkFANJ"
                  id="address-field"
                  name="address-field"
                  readonly=""
                  type="text"
                  value=""
                />
                <textarea
                  aria-label="location-story"
                  class="sc-crrsfI fwThyU"
                  id="location-story"
                  readonly=""
                  style="display: block;"
                />
              </div>
            </div>
          </div>
          <div
            aria-label="map-container"
            class="sc-dIUggk jmVEox"
            id="map-container"
          >
            <div
              class="sc-bdfBwQ cDyRiz"
            >
              <div
                class="sc-gsTCUz gbCiAz"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-fFubgz jfAIee"
    >
      <div
        class="sc-bkzZxe bzMjUo"
      >
        <div
          class="sc-idOhPF dyCZxh"
          id="addressListContainer"
        >
          <div
            class="sc-dQppl jjFnjm"
            id="address-list-container"
          >
            <div
              class="sc-bqyKva hKNAsX"
            >
              <div
                class="sc-kstrdz hrtAwC"
                id="geocoder-container"
              >
                <div
                  class="mapboxgl-ctrl-geocoder mapboxgl-ctrl"
                >
                  <svg
                    class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-search"
                    height="18"
                    viewBox="0 0 18 18"
                    width="18"
                    xml:space="preserve"
                  >
                    <path
                      d="M7.4 2.5c-2.7 0-4.9 2.2-4.9 4.9s2.2 4.9 4.9 4.9c1 0 1.8-.2 2.5-.8l3.7 3.7c.2.2.4.3.8.3.7 0 1.1-.4 1.1-1.1 0-.3-.1-.5-.3-.8L11.4 10c.4-.8.8-1.6.8-2.5.1-2.8-2.1-5-4.8-5zm0 1.6c1.8 0 3.2 1.4 3.2 3.2s-1.4 3.2-3.2 3.2-3.3-1.3-3.3-3.1 1.4-3.3 3.3-3.3z"
                    />
                  </svg>
                  <input
                    aria-label="Search"
                    class="mapboxgl-ctrl-geocoder--input"
                    placeholder="Search"
                    type="text"
                  />
                  <div
                    class="suggestions-wrapper"
                  >
                    <ul
                      class="suggestions"
                      style="display: none;"
                    />
                  </div>
                  <div
                    class="mapboxgl-ctrl-geocoder--pin-right"
                  >
                    <button
                      aria-label="Clear"
                      class="mapboxgl-ctrl-geocoder--button"
                    >
                      <svg
                        class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-close"
                        height="18"
                        viewBox="0 0 18 18"
                        width="18"
                        xml:space="preserve"
                      >
                        <path
                          d="M3.8 2.5c-.6 0-1.3.7-1.3 1.3 0 .3.2.7.5.8L7.2 9 3 13.2c-.3.3-.5.7-.5 1 0 .6.7 1.3 1.3 1.3.3 0 .7-.2 1-.5L9 10.8l4.2 4.2c.2.3.7.3 1 .3.6 0 1.3-.7 1.3-1.3 0-.3-.2-.7-.3-1l-4.4-4L15 4.6c.3-.2.5-.5.5-.8 0-.7-.7-1.3-1.3-1.3-.3 0-.7.2-1 .3L9 7.1 4.8 2.8c-.3-.1-.7-.3-1-.3z"
                        />
                      </svg>
                    </button>
                    <svg
                      class="mapboxgl-ctrl-geocoder--icon mapboxgl-ctrl-geocoder--icon-loading"
                      height="18"
                      viewBox="0 0 18 18"
                      width="18"
                      xml:space="preserve"
                    >
                      <path
                        d="M4.4 4.4l.8.8c2.1-2.1 5.5-2.1 7.6 0l.8-.8c-2.5-2.5-6.7-2.5-9.2 0z"
                        fill="#333"
                      />
                      <path
                        d="M12.8 12.9c-2.1 2.1-5.5 2.1-7.6 0-2.1-2.1-2.1-5.5 0-7.7l-.8-.8c-2.5 2.5-2.5 6.7 0 9.2s6.6 2.5 9.2 0 2.5-6.6 0-9.2l-.8.8c2.2 2.1 2.2 5.6 0 7.7z"
                        opacity=".1"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <br />
              <textarea
                aria-label="address-list-text-area"
                class="sc-hBEYos sc-fodVxV diFCsr kmUuBX"
                id="address-list-text-area"
                placeholder="Enter location story here"
              />
              <button
                aria-label="add-location-button"
                class="sc-hKgILt sc-iBPRYJ kYZczb gHMeds"
                id="add-location-button"
              >
                +
              </button>
            </div>
            <div
              class="sc-jrAGrp jmOtbB"
            >
              <div
                class="sc-kEjbxe ZjlhN"
              >
                <button
                  aria-label="left-arrow"
                  class="sc-dlfnbm ddKcEg"
                >
                  Left
                </button>
                <button
                  aria-label="right-arrow"
                  class="sc-dlfnbm ddKcEg"
                >
                  Right
                </button>
                <button
                  aria-label="trash-button"
                  class="sc-dlfnbm sc-eCssSg ddKcEg iQemPK"
                  id="trash-button"
                >
                  Trash
                </button>
                <button
                  aria-label="hide-text-button"
                  class="sc-dlfnbm ddKcEg"
                  id="hide-text-button"
                >
                  Hide/Show Text
                </button>
              </div>
              <input
                aria-label="address-field"
                class="sc-iqHYGH bkFANJ"
                id="address-field"
                name="address-field"
                readonly=""
                type="text"
                value=""
              />
              <textarea
                aria-label="location-story"
                class="sc-crrsfI fwThyU"
                id="location-story"
                readonly=""
                style="display: block;"
              />
            </div>
          </div>
        </div>
        <div
          aria-label="map-container"
          class="sc-dIUggk jmVEox"
          id="map-container"
        >
          <div
            class="sc-bdfBwQ cDyRiz"
          >
            <div
              class="sc-gsTCUz gbCiAz"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
